// semmle-io-node/out/file-reader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs-extra");
class FileReader {
    constructor(fd) {
        this._fd = fd;
    }
    dispose() {
        if (this._fd !== undefined) {
            fs.closeSync(this._fd);
            this._fd = undefined;
        }
    }
    get fd() {
        if (this._fd === undefined) {
            throw new Error('Object disposed.');
        }
        return this._fd;
    }
    readStream(start, end) {
        return new FileStreamReader(fs.createReadStream('', {
            fd: this.fd,
            start: start,
            end: end,
            autoClose: false
        }));
    }
    static async open(file) {
        const fd = await fs.open(file, 'r');
        return new FileReader(fd); // Take ownership
    }
}
exports.FileReader = FileReader;
class FileStreamReader {
    constructor(stream) {
        this._stream = stream;
    }
    [Symbol.asyncIterator]() {
        return this.stream[Symbol.asyncIterator]();
    }
    dispose() {
        if (this._stream !== undefined) {
            this._stream = undefined;
        }
    }
    get stream() {
        if (this._stream === undefined) {
            throw new Error('Object disposed.');
        }
        return this._stream;
    }
}
