// @playlyfe/gql-language-server/lib/providers/Hover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _vscodeLanguageserver = require("vscode-languageserver");

var _helpers = require("../utils/helpers");

class Hover {
  constructor({
    documents,
    gqlService
  }) {
    this.documents = void 0;
    this.gqlService = void 0;
    this.documents = documents;
    this.gqlService = gqlService;
  }

  provideHover(params) {
    const doc = this.documents.get(params.textDocument.uri);

    if (!doc) {
      // TODO: show warning
      return {
        contents: []
      };
    }

    const info = this.gqlService.getInfo({
      sourceText: doc.getText(),
      sourcePath: (0, _helpers.uriToFilePath)(params.textDocument.uri),
      position: (0, _helpers.lspPositionToGQLPosition)(params.position)
    });
    const result = {
      contents: []
    };
    info.forEach(item => {
      item.contents.forEach(content => {
        result.contents.push({
          language: 'graphql',
          value: content
        });
      });
    });
    return result;
  }

}

exports.default = Hover;
