// @apollo/federation/dist/composition/validate/preComposition/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var externalUsedOnBase_1 = require("./externalUsedOnBase");
Object.defineProperty(exports, "externalUsedOnBase", { enumerable: true, get: function () { return externalUsedOnBase_1.externalUsedOnBase; } });
var requiresUsedOnBase_1 = require("./requiresUsedOnBase");
Object.defineProperty(exports, "requiresUsedOnBase", { enumerable: true, get: function () { return requiresUsedOnBase_1.requiresUsedOnBase; } });
var keyFieldsMissingExternal_1 = require("./keyFieldsMissingExternal");
Object.defineProperty(exports, "keyFieldsMissingExternal", { enumerable: true, get: function () { return keyFieldsMissingExternal_1.keyFieldsMissingExternal; } });
var reservedFieldUsed_1 = require("./reservedFieldUsed");
Object.defineProperty(exports, "reservedFieldUsed", { enumerable: true, get: function () { return reservedFieldUsed_1.reservedFieldUsed; } });
var duplicateEnumOrScalar_1 = require("./duplicateEnumOrScalar");
Object.defineProperty(exports, "duplicateEnumOrScalar", { enumerable: true, get: function () { return duplicateEnumOrScalar_1.duplicateEnumOrScalar; } });
var duplicateEnumValue_1 = require("./duplicateEnumValue");
Object.defineProperty(exports, "duplicateEnumValue", { enumerable: true, get: function () { return duplicateEnumValue_1.duplicateEnumValue; } });
