// @divyenduz/graphql-language-service-server/dist/__mocks__/MockWatchmanClient.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 *  Copyright (c) Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the license found in the
 *  LICENSE file in the root directory of this source tree.
 *
 *  
 */

var MockWatchmanClient = function () {
  function MockWatchmanClient() {
    _classCallCheck(this, MockWatchmanClient);
  }

  MockWatchmanClient.prototype.checkVersion = function checkVersion() {
    return Promise.resolve();
  };

  MockWatchmanClient.prototype.listFiles = function listFiles(entryPath) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    return Promise.resolve([]);
  };

  MockWatchmanClient.prototype.runCommand = function runCommand() {
    return Promise.resolve();
  };

  MockWatchmanClient.prototype.watchProject = function watchProject(directoryPath) {
    return Promise.resolve({
      version: '',
      relative_path: '',
      watcher: '',
      watch: ''
    });
  };

  MockWatchmanClient.prototype.subscribe = function subscribe(entryPath, callback) {
    return Promise.resolve();
  };

  MockWatchmanClient.prototype.dispose = function dispose() {};

  return MockWatchmanClient;
}();

exports.default = MockWatchmanClient;
