// @divyenduz/graphql-language-service-server/dist/Logger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Logger = undefined;

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var _path = require('path');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } } /**
                                                                                                                                                           *  Copyright (c) Facebook, Inc.
                                                                                                                                                           *  All rights reserved.
                                                                                                                                                           *
                                                                                                                                                           *  This source code is licensed under the license found in the
                                                                                                                                                           *  LICENSE file in the root directory of this source tree.
                                                                                                                                                           *
                                                                                                                                                           *  
                                                                                                                                                           */

var SEVERITY = {
  ERROR: 'ERROR',
  WARNING: 'WARNING',
  INFO: 'INFO',
  DEBUG: 'DEBUG'
};

var Logger = exports.Logger = function () {
  function Logger() {
    _classCallCheck(this, Logger);

    var dir = (0, _path.join)(_os2.default.tmpdir(), 'graphql-language-service-logs');
    try {
      if (!_fs2.default.existsSync(dir)) {
        _fs2.default.mkdirSync(dir);
      }
    } catch (_) {
      // intentionally no-op. Don't block the language server even if
      // the necessary setup cannot be completed for logger.
    }

    this._logFilePath = (0, _path.join)(dir, 'graphql-language-service-log-' + _os2.default.userInfo().username + '-' + getDateString() + '.log');

    this._stream = null;
  }

  Logger.prototype.error = function error(message) {
    this._log(message, 'ERROR');
  };

  Logger.prototype.warn = function warn(message) {
    this._log(message, 'WARNING');
  };

  Logger.prototype.info = function info(message) {
    this._log(message, 'INFO');
  };

  Logger.prototype.log = function log(message) {
    this._log(message, 'DEBUG');
  };

  Logger.prototype._log = function _log(message) {
    var severityKey = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'DEBUG';

    var timestamp = new Date().toLocaleString(undefined);
    var severity = SEVERITY[severityKey];
    var pid = process.pid;

    var logMessage = timestamp + ' [' + severity + '] (pid: ' + pid + ') graphql-language-service-usage-logs: ' + message + '\n\n';
    // write to the file in tmpdir
    _fs2.default.appendFile(this._logFilePath, logMessage, function (error) {});
  };

  return Logger;
}();

// function getUnixTime() {
//   return new Date().getTime() / 1000;
// }

function getDateString() {
  var date = new Date();
  return date.getFullYear() + '-' + (date.getMonth() + 1) + '-' + date.getDate();
}
